### ISSUE TITLE
ci.bazel.io breakage: bazel-tests on windows, //src/test/shell/bazel:bazel_windows_example_test#test_cpp fails

### ISSUE THREAD

**Issue by damienmg** [bazelbuild/bazel#2178]
Log at:
http://ci.bazel.io/job/bazel-tests/BAZEL_VERSION=HEAD,PLATFORM_NAME=windows-x86_64/380/console

```
** test_cpp ********************************************************************
INFO: Reading 'startup' options from /c/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/_tmp/bazel_windows_example_test_1/bazelrc: --output_user_root=/c/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/_tmp/bazel_windows_example_test_1/root --host_javabase=C:/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/external/local_jdk
INFO: $TEST_TMPDIR defined: output root default is '/c/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/_tmp/bazel_windows_example_test_1'.
.............................................................................
____Loading package: examples/cpp
____Loading package: @bazel_tools//tools/cpp
____Loading package: @bazel_tools//tools/jdk
____Loading package: @local_config_xcode//
____Loading package: @local_jdk//
WARNING: C:/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/_tmp/bazel_windows_example_test_1/root/nrtg4st0/external/bazel_tools/tools/cpp/cc_configure.bzl:57:3: 
Auto-Configuration Warning: 'BAZEL_PYTHON' is not set, start looking for python in PATH.
.
WARNING: C:/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/_tmp/bazel_windows_example_test_1/root/nrtg4st0/external/bazel_tools/tools/cpp/cc_configure.bzl:57:3: 
Auto-Configuration Warning: Python found at C:/python_27_amd64/files/python.exe
.
WARNING: C:/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/_tmp/bazel_windows_example_test_1/root/nrtg4st0/external/bazel_tools/tools/cpp/cc_configure.bzl:57:3: 
Auto-Configuration Warning: 'BAZEL_VS' is not set, start looking for the latest Visual Studio installed.
.
ERROR: in target '//external:cc_toolchain': no such package '@local_config_cc//': Traceback (most recent call last):
	File "C:/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/_tmp/bazel_windows_example_test_1/root/nrtg4st0/external/bazel_tools/tools/cpp/cc_configure.bzl", line 520
		_find_vs_path(repository_ctx)
	File "C:/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/_tmp/bazel_windows_example_test_1/root/nrtg4st0/external/bazel_tools/tools/cpp/cc_configure.bzl", line 420, in _find_vs_path
		int(path_segments[-(3)].split(" ")[-..."))])
	File "C:/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/_tmp/bazel_windows_example_test_1/root/nrtg4st0/external/bazel_tools/tools/cpp/cc_configure.bzl", line 420, in int
		path_segments[-(3)]
Index out of range (index is -3, but sequence has 0 elements).
____Elapsed time: 10.363s
-- Test log: -----------------------------------------------------------
------------------------------------------------------------------------
test_cpp FAILED: Failed to build examples/cpp:hello-world .
/c/jenkins/workspace/bazel-tests/BAZEL_VERSION/HEAD/PLATFORM_NAME/windows-x86_64/src/test/shell/testenv.sh:447: in call to assert_build
/c/jenkins/workspace/bazel-tests/BAZEL_VERSION/HEAD/PLATFORM_NAME/windows-x86_64/src/test/shell/testenv.sh:453: in call to assert_build_output
C:/windows/TEMP/_bazel_SYSTEM/LhCiLHCO/execroot/windows-x86_64/bazel-out/local-fastbuild/bin/src/test/shell/bazel/bazel_windows_example_test:57: in call to test_cpp
FAILED: test_cpp
```

**Comment by laszlocsomor** [comment]
Looking.

**Comment by damienmg** [comment]
Note that I got that earlier with the previous build at http://ci.bazel.io/job/Gerrit-bazel-tests/216/ but wasn't present at HEAD so I suspect a flake

**Comment by meteorcloudy** [comment]
Looks like it's caused by my change https://github.com/bazelbuild/bazel/commit/325b1553d2469da862ebb894656eda69402511f8

But not sure why, I'll look into it.

**Comment by laszlocsomor** [comment]
Thanks @meteorcloudy .

@damienmg , the test is flaky, `test_java_test` failed for me on the first try, and there's a really insightful error message just 20 lines above the failure message or so:
```
==================== Test output for //examples/java-native/src/test/java/com/example/myproject:hello:
Error: Current working directory has a path longer than allowed for a
Win32 working directory.
Can't start native Windows application from here.
```

Meaning we need to focus on supporting long paths. That means we'll have to use the widechar version of Win32 API functions and prefix paths with `\\?\`.

**Comment by damienmg** [comment]
Deactivating the test until the underlying reason is fixed as there is no clear culprit.

**Comment by meteorcloudy** [comment]
The first appearance of this failure I can find is from http://ci.bazel.io/job/bazel-tests/BAZEL_VERSION=HEAD,PLATFORM_NAME=windows-x86_64/365/

So maybe it's a good start to investigate changes in that push.

**Comment by damienmg** [comment]
Removed the breakage label now the test is deactivated. please reactivate when fixed :)

**Comment by meteorcloudy** [comment]
I investigated on this, it turns out the [result.stdout](https://github.com/bazelbuild/bazel/blob/master/tools/cpp/cc_configure.bzl#L417) of `repository_ctx.execute` is sometimes incomplete and sometimes empty.

Having doubts on https://bazel-review.googlesource.com/#/c/7373/...

**Comment by meteorcloudy** [comment]
`bazel_windows_example_test` was originally failing due to VC detection was flaky on Windows, especially in shell test. This is no longer a problem since #2430 #2440 was fixed.

Currently, it is still failing because of two reasons:

1. workspace in shell test is getting too long (>260 chars), this can be workaround by setting `TEST_TMPDIR` to a short tmp directory.
```
Error: couldn't create '/c/tmp/_bazel_pcloudy/GYQDuV_U/execroot/bazel/_tmp/bazel_windows_example_test_1/root/install/912c8e92c0642cd6164a4c6f3bd7fed7.tmp.6488/_embedded_binaries/embedded_tools/third_party/checker_framework_dataflow/java/org/checkerframework/dataflow/analysis/AbstractValue.java': (error: 206): The filename or extension is too long.
```

2. `bazel_windows_example_test#test_java_test` fails when cleanup the workspace
Haven't found a good solution for this, I'm planning to disable this test case first and fix it later, follow up #2558

After these two issues, I think we can re-enable `bazel_windows_example_test`, which is important for getting test coverage for the major three languages.
